pkg/plugin/kms_v2_server.go

```diff
@@ -0,0 +1,144 @@
+		kvClient: kvClient,
+		reporter: metrics.NewStatsReporter(),
+	}, nil
+}
+
+// Status returns the health status of the KMS plugin
+func (s *KeyManagementServiceV2Server) Status(_ context.Context, _ *kmsv2.StatusRequest) (*kmsv2.StatusResponse, error) {
+	return &kmsv2.StatusResponse{
+		Version: version.V2APIVersion,
+		Healthz: "ok",
+		KeyId:   s.kvClient.GetAnnotations()["kms.plugin.io.keyid"],
+	}, nil
+}
+
+// Encrypt message
+func (s *KeyManagementServiceV2Server) Encrypt(ctx context.Context, request *kmsv2.EncryptRequest) (*kmsv2.EncryptResponse, error) {
+	klog.V(4).InfoS("encrypt request received", "id", request.Uid)
+	start := time.Now()
+
+	var err error
+	defer func() {
+		errors := ""
+		status := metrics.SuccessStatusTypeValue
+		if err != nil {
+			status = metrics.ErrorStatusTypeValue
+			errors = err.Error()
+		}
+		s.reporter.ReportRequest(ctx, metrics.EncryptOperationTypeValue, status, time.Since(start).Seconds(), errors)
+	}()
+
+	klog.V(2).Info("encrypt request started")
+	encryptResponse, err := s.kvClient.Encrypt(ctx, request.Plaintext)
+	if err != nil {
+		klog.ErrorS(err, "failed to encrypt")
+		return &kmsv2.EncryptResponse{}, err
+	}
+	klog.V(2).Info("encrypt request complete")
+
+	return &kmsv2.EncryptResponse{
+		Ciphertext:  encryptResponse.Ciphertext,
+		KeyId:       encryptResponse.KeyID,
+		Annotations: encryptResponse.Annotations,
+	}, nil
+}
+
+// Decrypt message
+func (s *KeyManagementServiceV2Server) Decrypt(ctx context.Context, request *kmsv2.DecryptRequest) (*kmsv2.DecryptResponse, error) {
+	klog.V(4).InfoS("decrypt request received", "id", request.Uid)
+	start := time.Now()
+
+	var err error
+	defer func() {
+		errors := ""
+		status := metrics.SuccessStatusTypeValue
+		if err != nil {
+			status = metrics.ErrorStatusTypeValue
+			errors = err.Error()
+		}
+		s.reporter.ReportRequest(ctx, metrics.DecryptOperationTypeValue, status, time.Since(start).Seconds(), errors)
+	}()
+
+	klog.V(2).Info("decrypt request started")
+
+	// validate keyID
+	annotations := s.kvClient.GetAnnotations()
+	if request.KeyId != annotations["kms.plugin.io.keyid"] {
+		err = fmt.Errorf("different keyID was used during encryption")
+		klog.ErrorS(err, "failed to decrypt", "keyID", request.KeyId)
+		return &kmsv2.DecryptResponse{}, err
+	}
+
+	// validate algorithm
+	if string(request.Annotations["kms.plugin.io.algorithm"]) != annotations["kms.plugin.io.algorithm"] {
+		err = fmt.Errorf("different algorithm was used during encryption")
+		klog.ErrorS(err, "failed to decrypt", "algorithm", string(request.Annotations["kms.plugin.io.algorithm"]))
+		return &kmsv2.DecryptResponse{}, err
+	}
+
+	// validate version
+	if string(request.Annotations["kms.plugin.io.version"]) != annotations["kms.plugin.io.version"] {
+		err = fmt.Errorf("different version was used during encryption")
+		klog.ErrorS(err, "failed to decrypt", "version", string(request.Annotations["kms.plugin.io.version"]))
+		return &kmsv2.DecryptResponse{}, err
+	}
```

Comment:
All of the annotation information is part of the keyvault client. Instead of `GetAnnotations` and compare here, can we add a function to the interface to validate annotations? That way none of these details are exposed outside of the keyvault package. 

That way we can get rid of this `GetAnnotations` method.